core/state/statedb.go

```diff
@@ -634,6 +635,47 @@ func (self *StateDB) GetRefund() uint64 {
 // Finalise finalises the state by removing the self destructed objects
 // and clears the journal as well as the refunds.
 func (s *StateDB) Finalise(deleteEmptyObjects bool) {
+	// Storage tries are independent, so update them in the background
+	updated := make([]*stateObject, 0, len(s.journal.dirties))
+
+	for addr := range s.journal.dirties {
+		stateObject, exist := s.stateObjects[addr]
+		if !exist {
+			// ripeMD is 'touched' at block 1714175, in tx 0x1237f737031e40bcde4a8b7e717b2d15e3ecadfe49bb1bbc71ee9deb09c6fcf2
+			// That tx goes out of gas, and although the notion of 'touched' does not exist there, the
+			// touch-event will still be recorded in the journal. Since ripeMD is a special snowflake,
+			// it will persist in the journal even though the journal is reverted. In this special circumstance,
+			// it may exist in `s.journal.dirties` but not in `s.stateObjects`.
+			// Thus, we can safely ignore it here
+			continue
+		}
+		if !stateObject.suicided && !(deleteEmptyObjects && stateObject.empty()) {
+			updated = append(updated, stateObject)
+		}
+	}
+	done := make(chan *stateObject, len(updated))
+
+	// Beside running the updates in the background, track the longest running one
+	// so we can report it as a storage update metric.
+	var (
+		start = time.Now()
+		end   time.Time
+		lock  sync.Mutex
+	)
+	for _, obj := range updated {
+		obj := obj // Closure, take care
+		workers.Schedule(func() {
+			obj.updateRoot(s.db)
+			if metrics.EnabledExpensive {
+				lock.Lock()
+				end = time.Now()
```

Comment:
Perhaps use an atomic here instead?

Reply:
Yeah, it's pretty ugly I agree. Trying to clean the code up a bit and get rid of some suboptimal calls in the mean time.